File: apps/releaf/ios/Releaf/Data/Auth/GoogleAuthClient.swift
/*
 * GoogleAuthClient.swift
 * Protocol + in-memory stub for Google Sign-In.
 *
 * Real implementation: swap `StubGoogleAuthClient` for a wrapper around
 * `GoogleSignIn-iOS` (https://github.com/google/GoogleSignIn-iOS).
 *
 * Scope required for Drive writes: `https://www.googleapis.com/auth/drive.file`
 * (Releaf can only see files it created).
 */

import Foundation

public struct GoogleAuthSession: Equatable, Sendable {
    public let userId: String
    public let email: String
    public let displayName: String?
    public let accessToken: String
    public let refreshToken: String?
    public let expiresAt: Date

    public init(
        userId: String,
        email: String,
        displayName: String? = nil,
        accessToken: String,
        refreshToken: String? = nil,
        expiresAt: Date
    ) {
        self.userId = userId
        self.email = email
        self.displayName = displayName
        self.accessToken = accessToken
        self.refreshToken = refreshToken
        self.expiresAt = expiresAt
    }
}

public enum GoogleAuthError: Error, Equatable {
    case cancelled
    case notImplemented
    case underlying(String)
}

public protocol GoogleAuthClient: AnyObject, Sendable {
    /// Launches the Google Sign-In flow and returns a session with Drive scope.
    func signIn() async throws -> GoogleAuthSession

    /// Refreshes the access token using the stored refresh token.
    func refresh(_ session: GoogleAuthSession) async throws -> GoogleAuthSession

    /// Revokes the token locally + remotely.
    func signOut() async
}

/// Default stub — lets the skeleton build + preview without the real SDK.
/// Returns a fake session after a short delay. Replace before shipping.
public final class StubGoogleAuthClient: GoogleAuthClient, @unchecked Sendable {
    public init() {}

    public func signIn() async throws -> GoogleAuthSession {
        try await Task.sleep(nanoseconds: 400_000_000)
        return GoogleAuthSession(
            userId: "stub-user",
            email: "you@example.com",
            displayName: "Preview User",
            accessToken: "stub-access-token",
            refreshToken: "stub-refresh-token",
            expiresAt: Date().addingTimeInterval(3_600)
        )
    }

    public func refresh(_ session: GoogleAuthSession) async throws -> GoogleAuthSession {
        return GoogleAuthSession(
            userId: session.userId,
            email: session.email,
            displayName: session.displayName,
            accessToken: session.accessToken,
            refreshToken: session.refreshToken,
            expiresAt: Date().addingTimeInterval(3_600)
        )
    }

    public func signOut() async {}
}

File: apps/releaf/android/app/src/main/java/app/releaf/mobile/auth/GoogleAuthClient.kt
/*
 * GoogleAuthClient.kt
 * Protocol + in-memory stub for Google Sign-In.
 *
 * Real implementation: wrap androidx.credentials.CredentialManager +
 * com.google.android.libraries.identity.googleid.GetGoogleIdOption for sign-in,
 * and com.google.android.gms.auth.api.identity.AuthorizationClient to request
 * the Drive scope.
 *
 * Scope required for Drive writes: `https://www.googleapis.com/auth/drive.file`.
 */

package app.releaf.mobile.auth

import kotlinx.coroutines.delay
import java.time.Instant

data class GoogleAuthSession(
    val userId: String,
    val email: String,
    val displayName: String?,
    val accessToken: String,
    val refreshToken: String?,
    val expiresAt: Instant,
)

sealed class GoogleAuthError(message: String) : Exception(message) {
    object Cancelled       : GoogleAuthError("Sign-in cancelled")
    object NotImplemented  : GoogleAuthError("Not implemented")
    class  Underlying(msg: String) : GoogleAuthError(msg)
}

interface GoogleAuthClient {
    suspend fun signIn(): GoogleAuthSession
    suspend fun refresh(session: GoogleAuthSession): GoogleAuthSession
    suspend fun signOut()
}

/**
 * Default stub — lets the skeleton build + preview without the real SDKs.
 * Returns a fake session after a short delay. Replace before shipping.
 */
class StubGoogleAuthClient : GoogleAuthClient {
    override suspend fun signIn(): GoogleAuthSession {
        delay(400)
        return GoogleAuthSession(
            userId = "stub-user",
            email = "you@example.com",
            displayName = "Preview User",
            accessToken = "stub-access-token",
            refreshToken = "stub-refresh-token",
            expiresAt = Instant.now().plusSeconds(3_600),
        )
    }

    override suspend fun refresh(session: GoogleAuthSession): GoogleAuthSession =
        session.copy(expiresAt = Instant.now().plusSeconds(3_600))

    override suspend fun signOut() = Unit
}
